scene/gui/foldable_container.cpp

```diff
@@ -0,0 +1,301 @@
+	}
+
+	if (theme_cache.panel_style.is_valid()) {
+		ms += theme_cache.panel_style->get_minimum_size();
+	}
+
+	ms.height += title_size.height;
+	return ms;
+}
+
+Vector<int> FoldableContainer::get_allowed_size_flags_horizontal() const {
+	Vector<int> flags;
+	flags.append(SIZE_FILL);
+	flags.append(SIZE_SHRINK_BEGIN);
+	flags.append(SIZE_SHRINK_CENTER);
+	flags.append(SIZE_SHRINK_END);
+	return flags;
+}
+
+Vector<int> FoldableContainer::get_allowed_size_flags_vertical() const {
+	Vector<int> flags;
+	flags.append(SIZE_FILL);
+	flags.append(SIZE_SHRINK_BEGIN);
+	flags.append(SIZE_SHRINK_CENTER);
+	flags.append(SIZE_SHRINK_END);
+	return flags;
+}
+
+void FoldableContainer::set_expanded(bool p_expanded) {
+	expanded = p_expanded;
+	for (int i = 0; i < get_child_count(); i++) {
+		Control *c = Object::cast_to<Control>(get_child(i));
+		if (!c || c->is_set_as_top_level()) {
+			continue;
+		}
+		c->set_visible(expanded);
+	}
+	queue_redraw();
+}
+
+bool FoldableContainer::is_expanded() const {
+	return expanded;
+}
+
+void FoldableContainer::set_title(String p_title) {
+	if (title == p_title) {
+		return;
+	}
+	title = p_title;
+	queue_redraw();
+}
+
+String FoldableContainer::get_title() const {
+	return title;
+}
+
+void FoldableContainer::set_title_alignment(HorizontalAlignment p_alignment) {
+	ERR_FAIL_INDEX((int)p_alignment, 3);
+	if (title_alignment == p_alignment) {
+		return;
+	}
+	title_alignment = p_alignment;
+	queue_redraw();
+}
+
+HorizontalAlignment FoldableContainer::get_title_alignment() const {
+	return title_alignment;
+}
+
+void FoldableContainer::gui_input(const Ref<InputEvent> &p_event) {
+	ERR_FAIL_COND(p_event.is_null());
+
+	Ref<InputEventMouseMotion> m = p_event;
+
+	if (m.is_valid()) {
+		if (m->get_position().y <= title_size.height) {
+			if (!is_hovering) {
+				is_hovering = true;
+				queue_redraw();
+			}
+		} else if (is_hovering) {
+			is_hovering = false;
+			queue_redraw();
+		}
+	}
+
+	if (has_focus() && p_event->is_action_pressed("ui_accept", false, true)) {
+		set_expanded(!expanded);
+		accept_event();
+		return;
+	}
+
+	Ref<InputEventMouseButton> b = p_event;
+
+	if (b.is_valid()) {
+		if (b->is_pressed() && b->get_button_index() == MouseButton::LEFT && b->get_position().y <= title_size.height) {
+			set_expanded(!expanded);
+			accept_event();
+			return;
+		}
+	}
+}
+
+void FoldableContainer::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_DRAW: {
+			RID ci = get_canvas_item();
+
+			Ref<Texture2D> icon = expanded ? theme_cache.arrow : theme_cache.arrow_collapsed;
+			Ref<StyleBox> title_style;
+			if (is_hovering) {
+				title_style = expanded ? theme_cache.title_hover_style : theme_cache.title_collapsed_hover_style;
+			} else {
+				title_style = expanded ? theme_cache.title_style : theme_cache.title_collapsed_style;
+			}
+
+			if (title_style.is_valid()) {
+				title_size = Point2(get_size().width, title_style->get_minimum_size().height);
```

Comment:
Need to initialize and set `title_size = Point2(get_size().width, 0)` before checking if the title_style is valid. I just need to know the case when a `StyleBox` will be invalid! because it's hard to make it null except from code.